### PR TITLE
Test 4.0.0

### LINKED ISSUE #42: Step 4.0: Basic End-to-End
This is the first 'end-to-end' test. 

* Simulation Component: GSM, GLEAM, EoR (powerlaw), gains, xtalk, reflections, RFI
* Simulators: `hera_sim`, `RIMEz`
* Pipeline Components: `redcal`, `smoothcal`,  `abscal`, `xRFI, `pspec`
* Depends on: #41 

## Why this test is required

This is the most basic end-to-end test: using basic components for every part of the simulation, but enough to include all analysis steps.

## Summary
A brief step-by-step description of the proposed test follows:

* Obtain FG and EoR visibilities in the same way as #16.
* Add gains in the same way as #16.
* Add systematics in the same way as #22 
* Add RFI in the same way as #41 
* Run full makeflow pipeline on the simulated data

## Simulation Details

* Freq. range:
* Channel width:
* Baseline/antenna configuration:
* Total integration time:
* Number of realisations:

## Criteria for Success

* P(k) matches known input to within 1%

### PR DESCRIPTION
**Pull request by steven-murray**
Resolves #42.

### CODE REVIEW

File: test-series/4/tools/sim_prep.py
Comment (steven-murray) [inline test-series/4/tools/sim_prep.py:38]: Don't need the `list` here.

**Comment by steven-murray** [comment]
There are `.pyc` files in the diff that are not getting git ignored. Should remove those.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:43]: Are you sure? I've had issues in the past with trying to access an individual element from a generator object, so I erred on the side of caution and made sure that it's a list rather than a generator.

**Comment by r-pascua** [comment]
> There are `.pyc` files in the diff that are not getting git ignored. Should remove those.

I think I've fixed this. I removed the `.pyc` files and added `*.pyc` to the gitignore. I hope this works, but I've had trouble getting the gitignore to do things I've wanted in the past.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:79]: They're not quite what `sim.add_xxx` does in the current version of `hera_sim`. I think in the version 1 branch, most of the tools are there, but the main issue right now is that multiplicative effects *do not* return the gains themselves, but rather the difference in the visibilities, and it's a whole lot of work to back out the gains from that data set.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:396]: I'm not too sure what the exact argument for this is to be honest--I'm only aware of the argument that there has been difficulty running redcal/abscal on simulation data before, so the safe route to take is to make sure all of the telescope-related information matches the observational metadata.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:477]: That's what I would've thought as well, but that doesn't seem to be the case. Another issue I had with it was that the `to_HERAData` function actually changes the class of the object that's passed to it, but I wasn't sure whether to raise an issue on the hera_cal github since it's in the legacy code section of the module.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:506]: I was not aware of capture groups! If you think it's better to use that, then I'll familiarize myself with their use and implement your suggestion.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:605]: That's much more concise! I'll make the change.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:615]: I'll check it out and see if it's worth pulling into here. The idea behind this function was that it would just be acting on one of the following files: `foregrounds.uvh5`, `eor.uvh5`, `sum.uvh5` (these are the compressed data sets for the full day of simulated data)

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:577]: Neat! I'll make the change.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:69]: I'm not sure I'm fully on board with this. I suppose the assumption I was making was that each polarization should have the same noise realization but different sky temperatures. If we think that the signal chains for the different polarizations are effectively uncorrelated, then I agree with you. That said, I still don't have a strong understanding of how to think about noise for an interferometer, so I'm leaning toward just taking your word for it.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:384]: Will do. The gist of it is that some baselines in the RIMEz simulation are conjugates of the baselines in the H1C array, so `blts` will be empty for those baselines (using `if blts:` breaks because `bool(blts)` throws an error for a nonempty `ndarray`).

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:396]: @jsdillon can you weigh in on this?

File: test-series/4/tools/sim_prep.py
Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:43]: Ah yeah, didn't see the access of the first element there. Really feels like there should be a better way to do this...

Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:43]: It would be nice if there was a simpler way to do it, but this is the best approach I can think of that will work with no prior knowledge of antenna numbers or which antennas actually have data associated with them.

Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:43]: Looks like the "correct" answer here is to do

```
next(ants for ants in sim.data.get_antpairs() if ants[0] == ants[1])
```

Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:43]: Ah, clever!

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:387]: I added a comment explaining the logic. I'll hold off on marking this as resolved until I get a confirmation that the comment is OK.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:408]: I added some clarifying comments, but I'm going to hold off on marking this as resolved--I'd like confirmation that the explanation I provided is OK.

File: test-series/4/tools/sim_prep.py
Comment (r-pascua) [reply test-series/4/tools/sim_prep.py:77]: I suppose I should also comment that the `add_reflections` and `add_xtalk` functions defined here go beyond what's capable even in the version 1 branch of hera_sim. A nice thing is that this has given me inspiration for some new functions to add to hera_sim to make this very easy to do. If you think it's worth getting the reflection/xtalk functions defined here pulled into hera_sim sooner than the version 1 release, then let's chat more about that via Slack?

Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:77]: Yup, let's do that :-) OK. Kind of annoying having to write all of this code for something that we fully intend hera_sim to be responsible for. Anyway, it's written now I guess! Yeah, when validation and Step 4 have been basically cracked, we should get back to v1.

File: test-series/4/tools/sim_prep.py
Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:408]: I guess I meant that it seems unintuitive to set the simulated "ideal-position" baselines to the "non-ideal" baselines of the data, since those baselines don't exactly match what was simulated. But IIRC @jsdillon was saying that the calibration pipeline never actually uses the real baselines, rather it just uses them to get the idealized positions and uses those. So in fact setting them to the non-ideal positions lets that assumption be tested a bit as well.

File: test-series/4/tools/sim_prep.py
Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:488]: wow... didn't even know that was possible! Really, `HERAData` should have a `.from_uvdata` classmethod.

File: test-series/4/tools/sim_prep.py
Comment (steven-murray) [reply test-series/4/tools/sim_prep.py:619]: Ahhh OK. Yeah I don't think the notebook function is for those files, it's for the chunked ones.

File: test-series/4/tools/prepare_abscal_model_files.pbs
Comment (steven-murray) [inline test-series/4/tools/prepare_abscal_model_files.pbs:31]: Is this right, should it be 1-7, or do we need a longer range here?

**Comment by steven-murray** [comment]
I'm happy (but can't technically approve since I made the PR...)

**Comment by r-pascua** [comment]
<p>Can do. This statement also needs updating to reflect the new EoR amplitude.</p><p> </p>

---

 <a href='https://app.reviewnb.com/HERA-Team/hera-validation/pull/54/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by r-pascua** [comment]
<p>Neat! One thing to note is that this function isn't going to work with the organization I decided to use (rather than living in <code>vis_path</code>, the files live in <code>vis_path / f'{jd_major}'</code>), and updating it will make it a little messy, or require that the <code>jd_major</code>  be specified.</p>

---

 <a href='https://app.reviewnb.com/HERA-Team/hera-validation/pull/54/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by r-pascua** [comment]
<p>My motivation for doing that was to meet @jsdillon's request that the simulation file metadata match the real data exactly. My understanding of the motivation behind the request is that there have been issues in the past with calibrating to simulations that were caused by metadata problems, but I'm not sure if that's exactly right. Either way, I do need to touch this up to reference the correct function in step 4.</p><p> </p>

---

 <a href='https://app.reviewnb.com/HERA-Team/hera-validation/pull/54/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by r-pascua** [comment]
<p>Ah, so this block is for writing the true gains to disk. It takes a fairly long time to run, so I commented it out for the latest run of the notebook. I can instead move this to a script or something, but I was hesitant to do that because it felt a little like I would be opening a can of worms, since I'd feel silly writing something that only writes <em>one</em> component to disk rather than any choice of component.</p>

---

 <a href='https://app.reviewnb.com/HERA-Team/hera-validation/pull/54/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by r-pascua** [comment]
<p>There was a point where there was a silent bug in the part of the code that handles adding time variation to the gains and so the bandpass gains were constant in time. That bug has since been fixed, so now the bandpass gains <em>do</em> vary in time.</p>

---

 <a href='https://app.reviewnb.com/HERA-Team/hera-validation/pull/54/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: test-series/4/tools/prepare_abscal_model_files.pbs
Comment (r-pascua) [reply test-series/4/tools/prepare_abscal_model_files.pbs:31]: The abscal model files are made to extend a bit past the simulated observation bounds. Originally the simulated observations spanned from 1-11 hours LST, so this choice gave us about an hour on each side. If I were to update the parameters, then I would maybe change the upper bound to 8 hours, but it really doesn't matter, up to a relatively small amount of disk space.

**Comment by r-pascua** [comment]
@steven-murray I think you may be looking at an outdated version of the notebook. (If you're clicking the ReviewNB link, then it might be showing you a very old version.) Here's a rendered version of the updated notebook: [https://nbviewer.jupyter.org/github/HERA-Team/hera-validation/blob/test-4.0.0/test-series/4/test-4.0.0a.ipynb](url)

All of the concerns raised in the previous comment should already be addressed. The only thing that I think needs to be addressed is that I only have a question mark put down next to the "Criteria" item in the description cell.

**Comment by steven-murray** [comment]
Ugh. OK, sorry for the noise. Yes, the current notebook looks excellent. For the criteria, I'd just say "N/A: test criteria appear in notebook 4.0.0c, in which power spectra are calculated".

**Comment by r-pascua** [comment]
Ok, great! Should I just make an "approve" review after pushing that change?

**Comment by steven-murray** [comment]
Yeah, sounds good!

**Review by r-pascua — APPROVED** [review]

**Pull request merged**
